pipeline/lib/pipeline.ml

```diff
@@ -231,26 +245,22 @@ let repositories = function
       in
       List.concat (List.concat repos)
 
-let exists ~conninfo repository =
-  let db = new Postgresql.connection ~conninfo () in
-  let exists = Benchmark.Db.exists db repository in
-  db#finish;
-  exists
+let exists ~db repository = Benchmark.Db.exists db repository
 
-let process_pipeline ~docker_config ~conninfo ~source () =
+let process_pipeline ~docker_config ~db ~source () =
   let run_args = Docker_config.run_args docker_config in
   Current.list_iter ~collapse_key:"pipeline"
     (module Repository)
     (fun repo ->
       let* repository = repo in
-      if exists ~conninfo repository then Current.ignore_value repo
-      else pipeline ~conninfo ~run_args repository)
+      if exists ~db repository then Current.ignore_value repo
+      else pipeline ~db ~run_args repository)
     (repositories source)
 
 let v ~current_config ~docker_config ~server:mode ~(source : Source.t) conninfo
     () =
-  Db_util.check_connection ~conninfo;
-  let pipeline = process_pipeline ~docker_config ~conninfo ~source in
+  let db = new Postgresql.connection ~conninfo () in
```

Comment:
I'm not sure that we want a long lived `db` connection, this can cause some issues because the database can restrict the number of active connections. In general it's better to have a short "connect/query/finish" even if it feels unoptimized! (We can look into pools if that really happens to be a bottleneck)